examples/command.ts

```diff
@@ -33,80 +33,115 @@
-  value: string
+  readonly _tag: 'Flag'
+  readonly value: string
 }
 
 export interface Named {
-  _tag: 'Named'
-  name: string
-  value: string
+  readonly _tag: 'Named'
+  readonly name: string
+  readonly value: string
 }
 
 export interface Positional {
-  _tag: 'Positional'
-  value: string
+  readonly _tag: 'Positional'
+  readonly value: string
+}
+
+export interface Args {
+  readonly flags: Array<string>
+  readonly named: Record<string, string>
+  readonly positional: Array<string>
 }
 
 export interface Ast {
-  command: string
-  source: string
-  args: {
-    flags: Record<string, true> | null
-    named: Record<string, string> | null
-    positional: Record<number, string> | null
-  }
+  readonly command: string
+  readonly source: string
+  readonly args: Args
 }
 
 // -------------------------------------------------------------------------------------
-// guards
+// instances
 // -------------------------------------------------------------------------------------
 
-const isFlag: Refinement<Argument, Flag> = (a): a is Flag => a._tag === 'Flag'
-
-const isNamed: Refinement<Argument, Named> = (a): a is Named => a._tag === 'Named'
-
-const isPositional: Refinement<Argument, Positional> = (a): a is Positional => a._tag === 'Positional'
+const monoidArgs: Monoid<Args> = getStructMonoid({
+  flags: A.getMonoid<string>(),
+  named: R.getMonoid(getLastSemigroup<string>()),
+  positional: A.getMonoid<string>()
+})
 
 // -------------------------------------------------------------------------------------
 // constructors
 // -------------------------------------------------------------------------------------
 
-export const Command = (value: string): Command => ({ _tag: 'Command', value })
-
 export const Flag = (value: string): Flag => ({ _tag: 'Flag', value })
 
 export const Named = (name: string, value: string): Named => ({ _tag: 'Named', name, value })
 
 export const Positional = (value: string): Positional => ({ _tag: 'Positional', value })
 
+export const FlagArg = (value: string): Args => ({
+  flags: [value],
+  named: R.empty,
+  positional: A.empty
+})
+
+export const NamedArg = (name: string, value: string): Args => ({
+  flags: A.empty,
+  named: { [name]: value },
+  positional: A.empty
+})
+
+export const PositionalArg = (value: string): Args => ({
+  flags: A.empty,
+  named: R.empty,
+  positional: [value]
+})
+
+// -------------------------------------------------------------------------------------
+// destructors
+// -------------------------------------------------------------------------------------
+
+export const fold = <R>(
+  onFlag: (value: string) => R,
+  onNamed: (name: string, value: string) => R,
+  onPositional: (value: string) => R
+) => (a: Argument): R => {
+  switch (a._tag) {
+    case 'Flag':
+      return onFlag(a.value)
+
+    case 'Named':
+      return onNamed(a.name, a.value)
+
+    case 'Positional':
+      return onPositional(a.value)
+
+    default:
+      return absurd<R>(a)
```

Comment:
I'm curious, why a `default` case using `absurd` here?

Reply:
I think it is mostly just force of habit on my part. Even though the TypeScript compiler provides exhaustive type checking for situations such as this, I always try to include a default case when a function is considered public (i.e. it is being exported). 

This way, if someone is using the lib in a pure JS environment and passes bad data to `fold`, they will be notified of their error rather than having to hunt down where the error occurred because the switch statement never executed.

However, for the purposes of the example if you would like to remove I am happy to do so!

Reply:
> if someone is using the lib in a pure JS environment

I see. Ok, LGTM, I'm going to merge, thank you!